### PR TITLE
Autoroute patterns localization

### PR DESCRIPTION
**Pull request by Skrypt**
This adds pattern localization to Orchard.Autoroute module. It's been tested against a clean install, 1.8.x and 1.9-int branch. For testing it you need to have at least 2 different culture set on your Orchard website and you need to activate Content Localization module, add a LocalizationPart to a custom content type. I've tested also import/export successfully.

Here is the configuration page : 
screencapture-localhost-30321-orchardlocal-admin-contenttypes-edit-page-1429480893312

Here is the result in an english page (see url) :
screencapture-localhost-30321-orchardlocal-admin-contents-edit-12-1429480909186

Here is the result in a french page (see url) :
screencapture-localhost-30321-orchardlocal-admin-contents-edit-13-1429481011218

I want to add also that the "Allow custom pattern" checkbox is useless right now. Would it be a feature that would be good to have eventually ?

### CODE REVIEW

**Comment by agriffard** [comment]
Please, use the correct bracket format.

**Comment by Skrypt** [comment]
Could you point me to the code design guidelines ?

**Comment by agriffard** [comment]
http://docs.orchardproject.net/Documentation/Code-conventions
http://stackoverflow.com/questions/3048800/how-can-i-set-visual-studio-to-use-kr-style-bracketing

File: src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs [hunk 1]
```diff
@@ -20,54 +24,85 @@ public class AutoroutePartDriver : ContentPartDriver<AutoroutePart> {
...
+            _cultureManager = cultureManager;
+            _httpContextAccessor = httpContextAccessor;
 
             T = NullLocalizer.Instance;
         }
 
         public Localizer T { get; set; }
 
-        protected override string Prefix { get { return "Autoroute"; }}
+        protected override string Prefix { get { return "Autoroute"; } }
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs]: I don't think this line is needed any more.

File: src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs [hunk 2]
```diff
@@ -20,54 +24,85 @@ public class AutoroutePartDriver : ContentPartDriver<AutoroutePart> {
...
+            if (part.Record.Id != 0) {
+                ContentItem contentItem = _contentManager.Get(part.Record.ContentItemRecord.Id);
+                if (contentItem != null) {
+                    dynamic contentItemForPart = contentItem;
+                    var ContentPart = contentItem.Parts.Where(x => x.PartDefinition.Name == "LocalizationPart").FirstOrDefault();
+
+                    // if the localizationpart is set on the content type we use it else we take the default website culture
+                    if (ContentPart.GetType().Name == "LocalizationPart") {
+                        itemCulture = contentItemForPart.LocalizationPart.Culture.Culture;
+                    }
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs]: Lines 65 -> 71

var aspect = contentItem.As<ILocalizationAspect>();

if (aspect != null){
   itemCulture = aspect.Culture;
}

File: src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs [hunk 3]
```diff
@@ -20,54 +24,85 @@ public class AutoroutePartDriver : ContentPartDriver<AutoroutePart> {
...
+
+            //if we are creating from a form post we use the form value for culture
+            HttpContextBase context = _httpContextAccessor.Current();
+            if (context.Request.Form["Localization.SelectedCulture"] != null) {
+                itemCulture = context.Request.Form["Localization.SelectedCulture"].ToString();
+            }
+
             // if the content type has no pattern for autoroute, then use a default one
-            if(!settings.Patterns.Any()) {
+            if (!settings.Patterns.Where(x => x.Culture == itemCulture).Any()) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs]: !settings.Patterns.Any(x => x.Culture == itemCulture)

File: src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs [hunk 4]
```diff
@@ -20,54 +24,85 @@ public class AutoroutePartDriver : ContentPartDriver<AutoroutePart> {
...
                 settings.AutomaticAdjustmentOnEdit = false;
-                settings.DefaultPatternIndex = 0;
-                settings.Patterns = new List<RoutePattern> {new RoutePattern {Name = "Title", Description = "my-title", Pattern = "{Content.Slug}"}};
+                settings.Patterns = new List<RoutePattern> { new RoutePattern { Name = "Title", Description = "my-title", Pattern = "{Content.Slug}", Culture = itemCulture } };
 
                 _notifier.Warning(T("No route patterns are currently defined for this Content Type. If you don't set one in the settings, a default one will be used."));
             }
 
+            // if the content type has no defaultPattern for autoroute, then use a default one
+            if (!settings.DefaultPatterns.Where(x => x.Culture == itemCulture).Any()) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Drivers/AutoroutePartDriver.cs]: !settings.DefaultPatterns.Any(x => x.Culture == itemCulture)

File: src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs [hunk 5]
```diff
@@ -27,30 +32,55 @@ public class AutorouteService : IAutorouteService {
...
             Logger = NullLogger.Instance;
-                T = NullLocalizer.Instance;
+            T = NullLocalizer.Instance;
         }
 
         public ILogger Logger { get; set; }
         public Localizer T { get; set; }
 
         public string GenerateAlias(AutoroutePart part) {
+            var itemCulture = _cultureManager.GetSiteCulture();
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs]: move this to after the check for part == null.

File: src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs [hunk 6]
```diff
@@ -27,30 +32,55 @@ public class AutorouteService : IAutorouteService {
...
+            if (part.Record.Id != 0) {
+                ContentItem contentItem = _contentManager.Get(part.Record.ContentItemRecord.Id);
+                if (contentItem != null) {
+                    dynamic contentItemForPart = contentItem;
+                    var ContentPart = contentItem.Parts.Where(x => x.PartDefinition.Name == "LocalizationPart").FirstOrDefault();
+
+                    // if the localizationpart is set on the content type we use it else we take the default website culture
+                    if (ContentPart.GetType().Name == "LocalizationPart") {
+                        itemCulture = contentItemForPart.LocalizationPart.Culture.Culture;
+                    }
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs]: use ILocalizationAspect

File: src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs [hunk 7]
```diff
@@ -110,15 +141,16 @@ public class AutorouteService : IAutorouteService {
...
 
-        public RoutePattern GetDefaultPattern(string contentType) {
+        public RoutePattern GetDefaultPattern(string contentType, string culture) {
             var settings = GetTypePartSettings(contentType).GetModel<AutorouteSettings>();
 
-            // return a default pattern if none is defined
-            if(settings.DefaultPatternIndex < settings.Patterns.Count) {
-                return settings.Patterns.ElementAt(settings.DefaultPatternIndex);    
+            // return a default pattern if set
+            if (settings.Patterns.Where(x => x.Culture == culture).Count() > 0) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Services/AutorouteService.cs]: why not use .Any() ?

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettings.cs [hunk 8]
```diff
@@ -39,18 +43,57 @@ public class AutorouteSettings {
...
+                if (_defaultPatterns == null) {
+                    _defaultPatterns = new JavaScriptSerializer().Deserialize<DefaultPattern[]>(DefaultPatternDefinitions).ToList();
+                }
+
+                //We split the values from the radio button returned values
+                int i = 0;
+                foreach (DefaultPattern defaultPattern in _defaultPatterns) {
+                    if (defaultPattern.Culture.Split('|').Count() > 1) {
+                        _defaultPatterns[i].PatternIndex = defaultPattern.Culture.Split('|').Last();
+                        _defaultPatterns[i].Culture = defaultPattern.Culture.Split('|').First();
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettings.cs]: should the splitter '|' be a constant?

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs [hunk 9]
```diff
@@ -25,8 +28,46 @@ public class AutorouteSettingsHooks : ContentDefinitionEditorEventsBase {
...
-            // add an empty pattern for the editor
-            settings.Patterns.Add(new RoutePattern());
+            //get cultures
+            settings.SiteCultures = _cultureManager.ListCultures();
+            //get default site culture
+            settings.DefaultSiteCulture = _cultureManager.GetSiteCulture();
+
+            //if a culture is not set on the token we set it to the default site culture for backward compatibility
+            //NOT SURE ABOUT THIS
+            if (settings.Patterns.Where(x => x.Culture == settings.DefaultSiteCulture).Count() == 0) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs]: !Any() ?

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs [hunk 10]
```diff
@@ -25,8 +28,46 @@ public class AutorouteSettingsHooks : ContentDefinitionEditorEventsBase {
...
+                            Culture = settings.DefaultSiteCulture
+                        });
+                    }
+                    current++;
+                }
+                //we add a new empty line for each culture
+                newPatterns.Add(new RoutePattern { Culture = culture, Name = null, Description = null, Pattern = null });
+
+                // if the content type has no defaultPattern for autoroute, then use a default one
+                if (!settings.DefaultPatterns.Where(x => x.Culture == culture).Any()) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs]: replace Where with Any()

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs [hunk 11]
```diff
@@ -25,8 +28,46 @@ public class AutorouteSettingsHooks : ContentDefinitionEditorEventsBase {
...
+                    settings.Patterns.Where(x => x.GetHashCode() == pattern.GetHashCode()).FirstOrDefault().Culture = settings.DefaultSiteCulture;
+                }
+            }
+
+            //Adding null Patterns for the ability to add another Pattern in the UI
+            List<RoutePattern> newPatterns = new List<RoutePattern>();
+            int current = 0;
+            foreach (string culture in settings.SiteCultures) {
+                foreach (RoutePattern routePattern in settings.Patterns.Where(x => x.Culture == culture)) {
+                    if (settings.Patterns.Where(x => x.Culture == culture).Any()) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs]: replace Where with Any()

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs [hunk 12]
```diff
@@ -39,31 +80,36 @@ public class AutorouteSettingsHooks : ContentDefinitionEditorEventsBase {
...
-                        Description = "my-title",
-                        Pattern = "{Content.Slug}"
-                    });
-
-                    _notifier.Warning(T("A default pattern has been added to AutoroutePart"));
+                //If there is no default pattern for each culture we set default ones
+                List<RoutePattern> newPatterns = new List<RoutePattern>();
+                int current = 0;
+                foreach (string culture in settings.SiteCultures) {
+                    if (settings.Patterns.Where(x => x.Culture == culture).Count() > 0) {
```
Comment (Jetski5822) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettingsEvents.cs]: .Any()

File: src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettings.cs (hunk 8 above)
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.Autoroute/Settings/AutorouteSettings.cs]: Why so mean ;)

**Comment by sebastienros** [comment]
Please target 1.x. It will definitely not make it for 1.9.

**Comment by Skrypt** [comment]
see pull request #5186 rebased on 1.x

**Pull request closed without merging**